### DISCUSSION TITLE
MongoDB

### DISCUSSION THREAD

**Discussion by barthofu** [barthofu/tscord#95] [Ideas]
The implementation of MongoDB in the template is something really important as it is used widely used (certainly not a good thing tho) by Discord bot developers.

As TSCord is using SQLite by default, it require some changes to switch it to mongodb, even if mikro-orm is abstracting a lot.

Theses changes has already been made in #16, but we don't really know yet how to apply these to the template. We've found two main solutions to this problem as far:

### 1. Documentation
Describe step by step the changes to make in the documentation.

<table>
<tr>
<td>Pros</td>
<td>Cons</td>
</tr>
<tr>
<td>
+ Really easy to "implement" <br>
</td>
<td>
- Not really user-friendly
</td>
</tr>
</table>

### 2. Templates
Having 2 different templates: the one using sqlite, and the one using mongodb.
We would then have to choose the one we want during the installation with the `tscord init bot` command.

> **Note**
> As the template is becoming more and more stable and close to its final version, it is a far more conceivable solution.

<table>
<tr>
<td>Pros</td>
<td>Cons</td>
</tr>
<tr>
<td>
+ More usre-friendly <br>
+ Would be the open gate to other variants of the template (e.g: a one using Prisma, etc)
</td>
<td>
- Each changes will have to be implemented in each variants <br>
- This will be a huge change involving a lot of efforts as it would certainly implies a switch to a Monorepo architecture
</td>
</tr>
</table>

---

We could also, in a first time, do the first solution as it is quite fast to write in the documentation, and then in the more long term implement the second solution.

**Comment by barthofu** [comment]
Ping @Mr-Artemus

**Comment by inBytes** [comment]
Hi there @barthofu !!

Just to note you that I tried to run the bot with the MongoDB configuration, and I faced a lot of problems.

I think is a little bit confuse [this part of the documentation](https://tscord.discbot.app/docs/bot/features/database/configuration) as you specify a tab for MongoDB, when in fact it is not possible to use it as is.

First, you have to add this line in "database.ts" for MongoDB section:

```
        dbName: process.env['DATABASE_NAME'],
``` 

I manage to make it work thanks to @Mr-Artemus commit #16.

The main problem was that this required dependency was completely missing from your package.json:

```
    "@mikro-orm/migrations-mongodb": "~5.6.16",
``` 

So it was impossible to execute this code in `node_modules\@mikro-orm\mongodb\MongoPlatform.js` (the line with the "require"):

```typescript
    getMigrator(em) {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const { Migrator } = require('@mikro-orm/migrations-mongodb');
        return this.config.getCachedService(Migrator, em);
    }
``` 
Beside of that, I think that the other main problem seems that in MongoDB (sorry, I'm very new to NoSQL) you are forced to use the "_id" field as Primary Key??? So you have to remove the @PrimaryKey decorator from every entity.

The only one that I had to do it by myself was the Pastebin one, that I guess it didn't exist when @Mr-Artemus did his job:

```typescript
import { Entity, PrimaryKey, Property, EntityRepositoryType } from "@mikro-orm/core"
import { EntityRepository } from "@mikro-orm/sqlite"

import { CustomBaseEntity } from "./BaseEntity"

// ===========================================
// ================= Entity ==================
// ===========================================

@Entity({ customRepository: () => PastebinRepository })
export class Pastebin extends CustomBaseEntity {

    [EntityRepositoryType]?: PastebinRepository

    @Property()
    editCode: string

    @Property()
    lifetime: number = -1

    @Property()
    createdAt: Date = new Date()
}

// ===========================================
// =========== Custom Repository =============
// ===========================================

export class PastebinRepository extends EntityRepository<Pastebin> { 

}
``` 

I tried to run the bot again but with 'better-sqlite' type in "database.ts", just to see if it was still compatible after the changes, but this error was thrown:

```
(ERROR) Unhandled rejection : Cannot redefine property: id
``` 

I don't know if this is something you can fix sometime in the future.

Thank you for your great work @barthofu !!

**Reply by barthofu** [reply]
Hello @inBytes 

Firstly, thanks for your investigation on this subject and for your financial support on the project!

Concerning MongoDB, it is true that we've totaly overlooked it for many times now and we decided update this branch asap to adapt it to the latest changes of the template. It'll be done not too soon after the `2.3` is released. You'll be credited because you provided much of the solution ;)

One thing I didn't understand; did you managed to make it work with Mongo at the end and if so, why did you went back to sqlite?

Btw, I don't know your project at all but NoSQL databases are not advised in the vast majority of cases, unless you know exactly what you are doing and it fills your specific needs.

One last thing; if you're fresh starting a new project with TSCord, we suggest you to wait for a couple days for the [v2.3](https://github.com/barthofu/tscord/pull/155) release as it will enhance and fix many quirks but will be quite a nightmare to migrate to sadly....

We hope you enjoy the project and encourage you to join the discord server and to make us feedbacks and/or directly contribute on it!

The Core Team.

**Comment by inBytes** [comment]
Hi again @barthofu !!

I really appreciate your work on this project and I hope to be able to contribute a fairer amount in the future.

Thank you very much for your intention to credit me, but I really think I don't deserve it, as most of the work was done by @Mr-Artemus.

I answer to your questions:
- I went back to sqlite (only momentarily) just to test if my code modifications were still compatible with any other database system, in order to propose you a PR for improve the project. But then... I discovered that the changes were only compatible with MongoDB.
- Yes, I went on purpose on the MongoDB path, because I'm pretty expert in relational databases (SQL Server, Oracle, MySQL) but I have zero experience in NoSQL, and I wanted to take advantage of this project to start learning about NoSQL at the same time. Later, if my experience with MongoDB is not good, I will switch back to MySQL without problems.
- And yes, I was looking at you out of the corner of my eye waiting for version 2.3 while learning to program my first interactions with the bot. I'll update to 2.3 when you get out the version, before going too far with my bot.

I'll get into your Discord after this message.

Thank you very much!

**Reply by barthofu** [reply]
I'm thrilled to hear your feedbacks on the project. Your willingness to contribute is greatly appreciated, and any future involvement is more than welcome!

Besides the fact that your efforts are valuable to the project, I understand your humility.

Thanks for the clarification on your database choices. Exploring MongoDB for NoSQL experience will be a fantastic journey for sure and we can't better know the pros/cons of a techno than by practicing it for sure.

It's fantastic to know that you're keeping an eye on the project's evolution and planning to update to version 2.3. We're here to support you in any way needed. Looking forward to having you in the Discord community!

Thank you very much for your dedication and engagement!
